### PR TITLE
Google Auth

### PR DESCRIPTION
**Pull request by UladBohdan**
* Google Authentication is implemented
* Image fetching is implemented
* Email/Password error messages has been reverted
* `config.json` file is updated as well as the one in Google Drive with real credentials

### CODE REVIEW

**Comment by UladBohdan** [comment]
[revert login api errors messages old format](https://trello.com/c/qR3LOjdF/100-revert-login-api-errors-messages-old-format)

**Comment by UladBohdan** [comment]
[to create apps in Social networks (fb, twitter, vk) both for social login and for sharing](https://trello.com/c/mAdDincK/142-to-create-apps-in-social-networks-fb-twitter-vk-both-for-social-login-and-for-sharing)

**Comment by UladBohdan** [comment]
[add social passport auth](https://trello.com/c/nTUT57Mw/35-add-social-passport-auth)

File: src/auth/index.js [hunk 1]
```diff
@@ -27,6 +27,9 @@ router.post('/register', requireFields('email', 'password', 'firstName'), (req,
     .catch(next);
 });
 
+router.get('/google', social.google.authenticate);
+router.get('/google/callback', social.google.callback, (req, res) => res.redirect('/'));
```
Comment (Drapegnik) [inline src/auth/index.js]: possible feature - redirect on page from `req.query.next` - page where user was, when clicked on `login`

File: src/auth/passport.js [hunk 2]
```diff
@@ -24,15 +27,15 @@ passport.use(
       User.findOne({ email, role: 'regular' })
         .then(result => {
           if (!result) {
-            throw new ValidationError({ password: 'Няправільны емэйл ці пароль' });
+            throw new ValidationError({ email: dictionary.be.badEmail });
```
Comment (Drapegnik) [inline src/auth/passport.js]: TODO: implement error codes, and move this dictionary on client dise

File: src/auth/passport.js [hunk 3]
```diff
@@ -79,5 +82,43 @@ const authenticate = (strategy, req, res, next) =>
     })(req, res, next);
   });
 
-export { authenticate };
+passport.use(
+  new GoogleStrategy(
+    {
+      clientID: config.social.google.clientId,
+      clientSecret: config.social.google.clientSecret,
+      callbackURL: 'http://localhost:8080/auth/google/callback',
```
Comment (Drapegnik) [inline src/auth/passport.js]: `http://localhost:8080`?

File: src/auth/passport.js [hunk 4]
```diff
@@ -79,5 +82,43 @@ const authenticate = (strategy, req, res, next) =>
...
+    },
+    (accessToken, refreshToken, profile, done) => {
+      const email = profile.emails[0].value;
+      User.findOne({ email })
+        .then(async user => {
+          if (user) {
+            return user;
+          }
+          const firstName = { unknown: profile.name && profile.name.givenName };
```
Comment (Drapegnik) [inline src/auth/passport.js]: `unknown` - like lang code?

File: src/auth/passport.js [hunk 5]
```diff
@@ -79,5 +82,43 @@ const authenticate = (strategy, req, res, next) =>
...
+          const lastName = { unknown: profile.name && profile.name.familyName };
+          const imageUrl = cutSizing(profile.photos && profile.photos[0].value);
```
Comment (Drapegnik) [inline src/auth/passport.js]: `profile.photos` or `profile.photos.length?`

File: src/auth/passport.js [hunk 6]
```diff
@@ -79,5 +82,43 @@ const authenticate = (strategy, req, res, next) =>
...
+    },
+    (accessToken, refreshToken, profile, done) => {
+      const email = profile.emails[0].value;
+      User.findOne({ email })
+        .then(async user => {
+          if (user) {
+            return user;
+          }
+          const firstName = { unknown: profile.name && profile.name.givenName };
+          const lastName = { unknown: profile.name && profile.name.familyName };
```
Comment (Drapegnik) [inline src/auth/passport.js]: const { givenName: firstName, familyName: lastName } = profile.name || {};

File: src/auth/passport.js (hunk 5 above)
Comment (Drapegnik) [inline src/auth/passport.js]: is't any other query params? we can made more generic method:

```js
const formatUrl = url => url && \\ cut all query params
```

and, please, place it not in `validation`, cause it not validator, but helper / formatter

File: src/auth/index.js (hunk 1 above)
Comment (UladBohdan) [reply src/auth/index.js]: that would be nice, agree

File: src/auth/passport.js (hunk 2 above)
Comment (UladBohdan) [reply src/auth/passport.js]: shall we introduce our own error coding system or to use an existing one? shall we send numbers or text?

File: src/auth/passport.js (hunk 3 above)
Comment (UladBohdan) [reply src/auth/passport.js]: temporary, I'll fix it before merging.

File: src/auth/passport.js (hunk 4 above)
Comment (UladBohdan) [reply src/auth/passport.js]: right.

File: src/auth/passport.js (hunk 5 above)
Comment (UladBohdan) [reply src/auth/passport.js]: seems to have the same behavior, isn't it?

File: src/auth/passport.js (hunk 6 above)
Comment (UladBohdan) [reply src/auth/passport.js]: not that easy, we also have language codes here

File: src/auth/passport.js (hunk 2 above)
Comment (Drapegnik) [reply src/auth/passport.js]: I imagined this like strings ids (`ERR_EMAIL`, `ERR_PASSW`). About existing system - we should research this question.

File: src/auth/passport.js (hunk 5 above)
Comment (Drapegnik) [reply src/auth/passport.js]: no:

File: src/auth/passport.js [hunk 7]
```diff
@@ -79,5 +82,43 @@ const authenticate = (strategy, req, res, next) =>
...
+      scope: [
+        'https://www.googleapis.com/auth/plus.login',
+        'https://www.googleapis.com/auth/plus.profile.emails.read',
+      ],
+    }),
+    callback: passport.authenticate('google', { failureRedirect: '/auth/login' }),
+  },
+};
+
+export { authenticate, social };
```
[the remark this answers is not in the document]
Comment (UladBohdan) [reply src/auth/passport.js]: I am not sure what do you mean by binding here. Where did you loose `local-login` then?

File: src/auth/passport.js [hunk 8]
```diff
@@ -68,16 +72,58 @@ passport.use(
...
+passport.use(
+  new GoogleStrategy(
+    {
+      clientID: config.social.google.clientId,
+      clientSecret: config.social.google.clientSecret,
+      callbackURL: '/auth/google/callback',
+    },
+    (accessToken, refreshToken, profile, done) => {
+      // profile here is as described in http://www.passportjs.org/docs/profile/
+      const email = profile.emails[0].value;
```
Comment (Drapegnik) [inline src/auth/passport.js]:
```js
(accessToken, refreshToken, { emails, name, photos }, done)
...
const [{ value: email }] = emails;
```

Comment (UladBohdan) [reply src/auth/passport.js]: done.

**Pull request merged**